datafusion/physical-expr/src/window/window_expr.rs

```diff
@@ -217,29 +221,73 @@ pub trait AggregateWindowExpr: WindowExpr {
 
     /// Calculates the window expression result for the given record batch.
     /// Assumes that `record_batch` belongs to a single partition.
+    #[allow(clippy::too_many_arguments)]
     fn get_result_column(
         &self,
         accumulator: &mut Box<dyn Accumulator>,
         record_batch: &RecordBatch,
+        most_recent_row: Option<&RecordBatch>,
         last_range: &mut Range<usize>,
         window_frame_ctx: &mut WindowFrameContext,
         mut idx: usize,
         not_end: bool,
     ) -> Result<ArrayRef> {
-        let values = self.evaluate_args(record_batch)?;
-        let order_bys = get_orderby_values(self.order_by_columns(record_batch)?);
+        let mut contains_most_recent_row = false;
+        let (values, order_bys) = if let Some(most_recent_row) = most_recent_row {
+            if is_record_batch_ahead(record_batch, most_recent_row, self.order_by())? {
+                contains_most_recent_row = true;
+                let batch = concat_batches(
```

Comment:
In this situation, the most recent information (called a watermark) is combined with the rest of the data. This method is a bit complex and affects how the algorithm works.

Rather than just using the current range of data to decide when the data is not safe to use, it's better to do a special safety check. This check is a more reliable way to figure out if the window data should not be used for emitting results